File: src/Java/algorithms/arrandstring/ZigZagConversion.java
package algorithms.arrandstring; 

import java.util.*; 

public class ZigZagConversion {
    /* Function: ZigZag in the vertical direction, turn string in each n chars, and return the result line by line. 
     * Solution: use mod operation. 
     */ 
    
    public String convert(String s, int nRows) {
        if (s == null || s.length() <= nRows || nRows <= 1) {
            return s; 
        }
        
        StringBuffer[] sbs = new StringBuffer[nRows]; 
        for (int i = 0; i < sbs.length; ++i) {
            sbs[i] = new StringBuffer(); 
        }
        
        int part = nRows + nRows - 2; 
        for (int i = 0; i < s.length(); ++i) {
            if (i % part < nRows) {
                sbs[i % part].append(s.charAt(i)); 
            } else {
                sbs[part - i % part].append(s.charAt(i)); 
            }
        }
        
        StringBuffer res = new StringBuffer(); 
        for (StringBuffer sb : sbs) {
            System.out.println(sb.toString()); 
            res.append(sb); 
        }
        
        return res.toString(); 
    }
}

File: src/CPP/src/leetcode/ZigZagConversion.cpp
/*
 * [Source] https://leetcode.com/problems/zigzag-conversion/
 * [Difficulty]: Easy
 * [Tag]: String
 */

#include <iostream>

using namespace std;

// [Solution]: divide the string by cycle
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/ZigZagConversion.java
 */

/* Java solution
public class Solution {
    public String convert(String s, int nRows) {
        if (s == null || s.length() == 0 || nRows <= 1) {
            return s; 
        }
        StringBuffer[] buffers = new StringBuffer[nRows]; 
        for (int i = 0; i < buffers.length; ++i) {
            buffers[i] = new StringBuffer(); 
        }
        int idx; 
        for (int i = 0; i < s.length(); ++i) {
            idx = i % (nRows + nRows - 2); 
            if (idx >= nRows) {
                idx = nRows + nRows - 2 - idx; 
            }
            buffers[idx].append(s.charAt(i)); 
        }
        for (int j = 1; j < buffers.length; ++j) {
            buffers[0].append(buffers[j]); 
        }
        return buffers[0].toString(); 
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
